### PR TITLE
feat(headless): add timezone to engines config, deprecate useLocalTime for buildDateRange

### PR DESCRIPTION
**Pull request by ThibodeauJF**
https://coveord.atlassian.net/browse/KIT-837

Should I also add it to the service APIs and the recommendations ?

### CODE REVIEW

File: packages/headless/src/utils/validate-payload.ts [hunk 1]
```diff
@@ -13,6 +13,11 @@ export const requiredNonEmptyString = new StringValue({
   emptyAllowed: false,
 });
 
+export const nonEmptyString = new StringValue({
```
Comment (ThibodeauJF) [inline packages/headless/src/utils/validate-payload.ts:16]: This definition was used often, so I replaced with this everywhere I could see

File: packages/atomic/src/components.d.ts [hunk 2]
```diff
@@ -5,1773 +5,21 @@
  * It contains typing information for all components that exist in this project.
  */
 import { HTMLStencilElement, JSXBase } from "@stencil/core/internal";
-import { CategoryFacetSortCriterion, DateFilter, DateFilterState, FacetSortCriterion, LogLevel, NumericFilter, NumericFilterState, RangeFacetRangeAlgorithm, RangeFacetSortCriterion, RelativeDateUnit, Result, ResultTemplate, ResultTemplateCondition, SearchEngine } from "@coveo/headless";
```
Comment (ThibodeauJF) [inline packages/atomic/src/components.d.ts:8]: Hmmm I'll fix this

File: packages/headless/src/app/search-engine/search-engine-configuration.ts [hunk 3]
```diff
@@ -42,6 +42,14 @@ export interface SearchConfigurationOptions {
    *  Moreover, this information can be referred to in query expressions and QPL statements by using the $locale object.
    */
   locale?: string;
+  /**
+   * The [tz database](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones) identifier of the time zone to use to correctly interpret dates in the query expression, facets and result items.
```
Comment (dlutzCoveo) [inline packages/headless/src/app/search-engine/search-engine-configuration.ts:46]:
```suggestion
   * The [tz database](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones) identifier of the time zone to use to correctly interpret dates in the query expression, facets, and result items.
```

**Review by dlutzCoveo — APPROVED** [review]

File: packages/headless/src/utils/validate-payload.ts (hunk 1 above)
Comment (olamothe) [reply packages/headless/src/utils/validate-payload.ts:16]: Just a thought: Is there a possibility that there's some stateful properties in a`StringValue` instance from bueno ?

Should it be a function that returns a `new StringValue({..})` instead ?

File: packages/headless/src/controllers/facets/range-facet/date-facet/date-range.test.ts [hunk 4]
```diff
@@ -1,3 +1,5 @@
+import dayjs from 'dayjs';
```
Comment (olamothe) [inline packages/headless/src/controllers/facets/range-facet/date-facet/date-range.test.ts:1]: The changes in this file seems unrelated. 
What's the story behind those ?

**Review by olamothe — APPROVED** [review]
Yes, I think we should do the change for all use cases.

File: packages/headless/src/controllers/facets/range-facet/date-facet/date-range.test.ts (hunk 4 above)
Comment (ThibodeauJF) [reply packages/headless/src/controllers/facets/range-facet/date-facet/date-range.test.ts:1]: Since all those tests were done with useLocalTime = false, the values were all UTC. This causes a problem with hardcoding expected values, so I used dayjs everywhere

File: packages/headless/src/utils/validate-payload.ts (hunk 1 above)
Comment (ThibodeauJF) [reply packages/headless/src/utils/validate-payload.ts:16]: I don't think it's a possibility since it's a definition and shouldn't be modified, we also have had requiredNonEmptyString / requiredEmptyAllowedString for the longest time

File: packages/headless/src/controllers/facets/range-facet/date-facet/date-range.test.ts (hunk 4 above)
Comment (olamothe) [reply packages/headless/src/controllers/facets/range-facet/date-facet/date-range.test.ts:1]: Ah, makes sense. Thanks

**Review by btaillon — APPROVED** [review]

**Comment by ThibodeauJF** [comment]
@olamothe 
Looked it up & timezone is an available parameter for the following requests:
- / (search & recommendations)
- /querySuggest
- /plan

Not available for:
- /facet (facet search)
- /html
- /classify (case assist)
- /documents/suggest (case assist)

**Review by samisayegh — APPROVED** [review]

**Review by nbuissoncoveo — APPROVED** [review]
